src/gridfs/upload.rs

```diff
@@ -376,7 +377,7 @@ impl Drop for GridFsUploadStream {
         if !matches!(self.state, State::Closed) {
             let chunks = self.bucket.chunks().clone();
             let id = self.id.clone();
-            runtime::execute(async move {
+            self.drop_token.spawn(async move {
```

Comment:
I wrote the following test because I was curious about what "bad" usage of `shutdown` would cause in a situation like this:

```
#[cfg_attr(feature = "tokio-runtime", tokio::test)]
#[cfg_attr(feature = "async-std-runtime", async_std::test)]
async fn stream_shutdown() {
    use futures_util::AsyncWriteExt;

    let client = TestClient::new().await.into_client();

    let db = client.database("db");
    let bucket = db.gridfs_bucket(None);
    let mut stream = bucket.open_upload_stream("test", None);

    let buf = vec![0u8; 100];
    stream.write(&buf[..]).await.unwrap();

    client.shutdown().await;

    stream.close().await.unwrap();
}
```
The result was a hang, due to the fact that `shutdown` will wait either for a cleanup future to be registered for the stream or for its sender to be dropped, neither of which can happen if `stream` is used after `shutdown` is called. IIUC this relates to your first bullet point in the overall comment for this PR (that correct ordering is on the user), but I do think that, if a user fumbles this in a situation in which guaranteeing order is somehow difficult, diagnosing a hang like this will be challenging.

I wonder if we can somehow flip the direction of these messages: the stream passes its cleanup future to `register_async_drop` so that the client has a handle to it immediately, `register_async_drop` returns a receiver that will get an ack if `shutdown` is called before `GridFsUploadStream::drop`, and `drop` uses the state of the receiver to determine whether it should do the cleanup itself. Then `shutdown` will always be guaranteed to complete and I *think* the call to `stream.close()` would return an error that the client has already been shut down. (This was just a quick idea so it's very possible I'm missing some complications that would make it infeasible.)

Also, will this drop code be run if a user doesn't call either shutdown method?

Reply:
Registering the cleanup future at construction with `register_async_drop` runs into the same shared-state issue that my initial approach did - that future will need to co-own some subset of the internal state of the stream along with the `Stream` struct itself, which in turn means access only through a mutex, which in turn means nothing can return a plain `&ref`.

Ideally the behavior would be what you laid out: `shutdown` or `drop` would both check to see if cleanup had already happened, and if not do it immediately.  That's what I spent most of the time on this ticket trying to build, but I haven't been able to find a way to make that work with the Rust ownership model and our current API.

The current state of this PR is my stab at a second-best sort of solution.  It starts with the assumption that ownership of resource state will lie solely with the associated struct, which rules out any kind of pre-emptive cleanup when `shutdown` executes.  Waiting for outstanding resources to be cleaned up at least means we can give a clear guarantee to the user: "When `shutdown()` has completed, everything is tidily closed out."

I agree that a hang is not a great user experience.  My hypothesis is that the "natural" way to use this library is unlikely to run into that, and where the user is *expecting* `shutdown` to guarantee everything is tidied up a hang is at least a clear signal that cleanup isn't happening.  As an escape hatch, `shutdown_immediate` exists for situations where that's not possible.  I've gone back and updated the comment on `shutdown` to be more clear about what's needed there. 

> Also, will this drop code be run if a user doesn't call either shutdown method?

Yup, the cleanup code runs as soon as `AsyncDropToken::spawn` runs (and the task scheduler decides to pick up that task).  `shutdown` doesn't trigger that execution, it just waits for it.

Reply:
Ah that makes a lot more sense. I forgot that the task is still executing in the background regardless of whether `shutdown` is called.

Yeah, I think it would be nice to revisit some of this when we can make breaking changes -- can you file a ticket with a 3.0 fix version so we remember to think about it more then?